adoptingerlang.org

```diff
@@ -3234,20 +3257,104 @@ envFrom:
 #+end_admonition
 
 [TODO: show how to have a crashdump saved to a persistent volume]
 
+**** Init Containers
+
+Our release of =service_discovery= depends on a Postgres database for storing state, to ensure the database is properly setup before running our containers we will use an [[https://kubernetes.io/docs/concepts/workloads/pods/init-containers/][Init Container]]. Each =Init Container= is a container that runs to completion before any of the main containers in a Pod are run. The database migration tool =flyway= we are using for running migrations is used to validate the database is up to date before allowing the main containers to run:
+
+#+BEGIN_SRC yaml
+volumes:
+- name: migrations
+  emptyDir:
+    medium: Memory
+
+initContainers:
+- name: flyway
+  image: boxfuse/flyway:5.2.4
+  args:
+  - "-url=jdbc:postgresql://$(POSTGRES_SERVICE):5432/discovery"
+  - -user=discovery
+  - -password=password
+  - -connectRetries=60
+  - validate
+  volumeMounts:
+  - name: migrations
+    mountPath: /flyway/sql
+
+- name: service-discovery-sql
+  image: service_discovery
+  command: ["/bin/sh"]
+  args: ["-c", "cp /opt/service_discovery/sql/* /flyway/sql"]
+  volumeMounts:
+    - name: migrations
+      mountPath: /flyway/sql
+#+END_SRC
+
+The volume =migrations= exists only in memory and is used to copy the current SQL migration files from the release image to a shared directory that the =flyway= container will use to run validation. The =emptyDir= type of volume with =medium= =memory= is used because we are not using this volume to persist any data, only to share between two =Init Contianers=. If =medium: Memory= was not included the =emptyDir= is created on the host file system but it is still empty initially for each new =Pod= and is removed when the =Pod= is deleted.
+
+**** Readiness and Liveness Probes
+
+Kubernetes has multiple configurable probes to know when to restart a container and when a container is in a state it can accept traffic. The main two are the liveness and readiness probes, which handle container restart and whether traffic can be sent respectively. Additionally, a startup probe can be used to delay when the liveness and readiness probe start being checked.
+
+Probes are defined as either a command to execute in the container, a HTTP =GET= request to make to a port and path or an attempt to open a TCP connection to a specific port.
+
+A =livenessProbe= is not necessarily needed. If the process in the container crashes or stop for any reason it will be restarted. The probe ensures that if our node does not crash but for some reason has deadlocked and can not respond to requests it will also be restarted.
```

Comment:
```suggestion
A =livenessProbe= is optional; if the process in the container crashes or stops for any reason, it will be restarted. The probe ensures that if our node does not crash but for some reason has deadlocked and can not respond to requests, it will also be restarted.
```